Automattic/newspack-blocks
# fix(modal-checkout): optimize iframe load and refactor anonymous cart generation

### All Submissions:






### Changes proposed in this Pull Request:

This PR addresses a few different issues:

Introduces a new strategy to detect iframe load so it can be ready sooner. The `load` event dispatch waits for render-blocking assets to load before firing. The new strategy watches for the iframe content document to trigger when the DOM is ready.

The dequeuing strategy is now reversed so that only whitelisted assets can be enqueued. This prevents unnecessary assets from loading.

#### Cart generation via ajax

To handle an edge case in which carts are getting wiped between an account registration and the modal checkout render, this PR changes the strategy for anonymous readers.

Carts should now be generated in parallel to the auth/account creation flow. This allows for the cart session to be created before the authentication, which, once authenticated, Woo should handle the transition without issues.

Because of this new strategy, readers signing in via link should also be able to restore the checkout flow with the existing cart. The pending checkout logic has been simplified here and in https://github.com/Automattic/newspack-plugin/pull/3471.

### How to test the changes in this Pull Request:

1. Checkout this branch and https://github.com/Automattic/newspack-plugin/pull/3471
2. In a fresh session, start the checkout button flow with a subscription product
3. Register a new reader and confirm you are able to purchase without issues
4. In another fresh session, start the checkout button flow with the same product
5. This time, enter an existing email and click to sign in via link
6. Paste the link in your session and confirm you are authenticated and able to proceed with the checkout
7. Repeat steps 2-6 but through the Donate block

### Other information:
